File: apps/api/src/lib/events.ts
import type { Redis } from "ioredis";
import { log } from "./logger";

let redisInstance: Redis | null = null;

export const initEventBus = (redis: Redis): void => {
  redisInstance = redis;
};

export const getEventRedis = (): Redis | null => redisInstance;

export const publishEvent = async (
  type: string,
  data: unknown
): Promise<void> => {
  if (!redisInstance) return;
  const event = { data, timestamp: new Date().toISOString(), type };
  await redisInstance
    .publish("raven:events", JSON.stringify(event))
    .catch((err) => {
      log.error("Failed to publish event", err, { type });
    });
};

File: apps/api/src/lib/events.py
import json
from typing import Any

from src.redis import get_redis

CHANNEL = "raven:events"


async def publish(event_type: str, data: dict[str, Any]) -> None:
    redis = get_redis()
    payload = json.dumps({"type": event_type, "data": data})
    await redis.publish(CHANNEL, payload)


async def subscribe():
    redis = get_redis()
    pubsub = redis.pubsub()
    await pubsub.subscribe(CHANNEL)
    return pubsub
